lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +182,212 @@ class Loader {
-        return require(builtinName);
+    for (let i = count - 1; i > -1; i--) {
+      const preloadCode = this.#globalPreloaders[i]();
+      if (preloadCode === null) return;
+
+      if (typeof preloadCode !== 'string') {
+        throw new ERR_INVALID_RETURN_VALUE(
+          'string',
+          'loader globalPreloadCode',
+          preloadCode,
+        );
       }
-      throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
-    });
+      const { compileFunction } = require('vm');
+      const preloadInit = compileFunction(
+        preloadCode,
+        ['getBuiltin'],
+        {
+          filename: '<preload>',
+        }
+      );
+      const { NativeModule } = require('internal/bootstrap/loaders');
+
+      preloadInit.call(globalThis, (builtinName) => {
+        if (NativeModule.canBeRequiredByUsers(builtinName)) {
+          return require(builtinName);
+        }
+        throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
+      });
+    }
   }
 
-  async getModuleJob(specifier, parentURL) {
-    const url = await this.resolve(specifier, parentURL);
-    const format = await this.getFormat(url);
-    let job = this.moduleMap.get(url);
-    // CommonJS will set functions for lazy job evaluation.
-    if (typeof job === 'function')
-      this.moduleMap.set(url, job = job());
-    if (job !== undefined)
-      return job;
+  /**
+   * Resolve the location of the module.
+   *
+   * The internals of this WILL change when chaining is implemented,
+   * depending on the resolution/consensus from #36954
+   * @param {*} originalSpecifier - The specified URL path of the module to be
+   * resolved
+   * @param {String} parentURL - The URL path of the module's parent
+   * @returns {{ url: String }}
+   */
+  async resolve(originalSpecifier, parentURL) {
+    const conditions = DEFAULT_CONDITIONS;
 
-    if (!translators.has(format))
-      throw new ERR_UNKNOWN_MODULE_FORMAT(format);
+    const [ defaultResolver, customResolver ] = this.#resolvers;
 
-    const loaderInstance = translators.get(format);
+    const resolver = customResolver || defaultResolver;
+    const resolution = resolver(
+      originalSpecifier,
+      {
+        conditions,
+        parentURL,
+      },
+      defaultResolver,
+    );
 
-    const inspectBrk = parentURL === undefined &&
-        format === 'module' && getOptionValue('--inspect-brk');
-    job = new ModuleJob(this, url, loaderInstance, parentURL === undefined,
-                        inspectBrk);
-    this.moduleMap.set(url, job);
-    return job;
+    if (typeof resolution !== 'object') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'object',
+        'loader resolve',
+        resolution,
+      );
+    }
+
+    const { url } = resolution;
+
+    if (typeof url !== 'string') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'string',
+        'loader resolve',
+        'url',
+        url,
+      );
+    }
+
+    return {
+      url,
```

Comment:
It would be nice to be able to still return a format from the resolve hook to hint the load phase. Often the format information is discovered along with the resolve algorithm in the first place.

Reply:
I could see that being useful and don't see any harm, but I believe that's not what I was supposed to do? (specifically: [Geoffrey's comment in #36396 on 12 Jan 2020](https://github.com/nodejs/node/issues/36396#issuecomment-759072459)) I'd be happy to chuck it in if desired. That would also then make load's `context` useful (it's currently empty).

Reply:
//cc @GeoffreyBooth

Reply:
Agreed in last week's meeting to support as described here. Implemented in: b46eb60